RefineDetectedMarkers fails if cameraMatrix is given as a param

### System Information

OpenCV python version: 4.8.0
Operating System / Platform: Windows 10
Python version: 3.9.16

### Detailed description

`refineDetectedMarkers `method of `ArucoDetector `fails if optional `cameraMatrix `param is given. 

When running following line, 
```
corners, ids, rejected, _ = aruco_detector.refineDetectedMarkers(
    image=grayscale_image,
    board=charucoboard,
    detectedCorners=aruco_corners,
    detectedIds=aruco_ids,
    rejectedCorners=aruco_rejected,
    cameraMatrix=camera_matrix,
    distCoeffs=dist_coeffs,
)
```
An assertion error is raised: 
```
    corners, ids, rejected, _ = aruco_detector.refineDetectedMarkers(
cv2.error: OpenCV(4.8.0) D:\bld\libopencv_1690022693676\work\modules\objdetect\src\aruco\aruco_board.cpp:460: error: (-215:Assertion failed) (int)detectedCharucoVecMat[i].total() * detectedCharucoVecMat[i].channels() == 2 in function 'cv::aruco::CharucoBoardImpl::matchImagePoints'
```
No error is raised if no `cameraMatrix` is specified. 


### Steps to reproduce

```python 
import imageio.v3 as iio
import numpy as np
import cv2

grayscale_image = iio.imread( r"C:\Users\jnicks\Documents\repositories\camera-characterization\gray_scale_image.png")
camera_matrix = np.array(
    [
        [1.08444638e03, 0.00000000e00, 6.99703945e02],
        [0.00000000e00, 1.08545749e03, 5.24978509e02],
        [0.00000000e00, 0.00000000e00, 1.00000000e00],
    ]
)
dist_coeffs = np.array([0.01183206, 0.01096358, 0.0, 0.0])

charucoboard = cv2.aruco.CharucoBoard(
    (11, 8),
    35,
    26,
    cv2.aruco.getPredefinedDictionary(cv2.aruco.DICT_4X4_50),
)
charucoboard.setLegacyPattern(True)
aruco_detector = cv2.aruco.ArucoDetector(
    dictionary=charucoboard.getDictionary(), detectorParams=cv2.aruco.DetectorParameters()
)
aruco_corners, aruco_ids, aruco_rejected = aruco_detector.detectMarkers(image=grayscale_image)
# following line works
# corners, ids, rejected, _ = aruco_detector.refineDetectedMarkers(
#     image=grayscale_image,
#     board=charucoboard,
#     detectedCorners=aruco_corners,
#     detectedIds=aruco_ids,
#     rejectedCorners=aruco_rejected,
#     distCoeffs=dist_coeffs,
# )
# following line fails
corners, ids, rejected, _ = aruco_detector.refineDetectedMarkers(
    image=grayscale_image,
    board=charucoboard,
    detectedCorners=aruco_corners,
    detectedIds=aruco_ids,
    rejectedCorners=aruco_rejected,
    cameraMatrix=camera_matrix,
    distCoeffs=dist_coeffs,
)

```
Following image is used as input: ![gray_scale_image](https://github.com/opencv/opencv/assets/135614855/5b57639b-0e25-4003-b071-3cf83638f84e)


### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

--- a/modules/objdetect/misc/python/test/test_objdetect_aruco.py
+++ b/modules/objdetect/misc/python/test/test_objdetect_aruco.py
@@ -186,6 +186,39 @@ class aruco_objdetect_test(NewOpenCVTests):
         self.assertEqual((1, 4, 2), refine_corners[0].shape)
         np.testing.assert_array_equal(corners, refine_corners)
 
+    def test_charuco_refine(self):
+        aruco_dict = cv.aruco.getPredefinedDictionary(cv.aruco.DICT_6X6_50)
+        board_size = (3, 4)
+        board = cv.aruco.CharucoBoard(board_size, 1., .7, aruco_dict)
+        aruco_detector = cv.aruco.ArucoDetector(aruco_dict)
+        charuco_detector = cv.aruco.CharucoDetector(board)
+        cell_size = 100
+        image = board.generateImage((cell_size*board_size[0], cell_size*board_size[1]))
+        camera = np.array([[1, 0, 0.5],
+                           [0, 1, 0.5],
+                           [0, 0, 1]])
+        dist = np.array([0, 0, 0, 0, 0], dtype=np.float32).reshape(1, -1)
+
+        # generate gold corners of the ArUco markers for the test
+        gold_corners = np.array(board.getObjPoints())[:, :, 0:2]*cell_size
+
+        # detect corners
+        markerCorners, markerIds, _ = aruco_detector.detectMarkers(image)
+
+        # test refine
+        rejected = [markerCorners[-1]]
+        markerCorners, markerIds = markerCorners[:-1], markerIds[:-1]
+        markerCorners, markerIds, _, _ = aruco_detector.refineDetectedMarkers(image, board, markerCorners, markerIds,
+                                                                              rejected, cameraMatrix=camera, distCoeffs=dist)
+
+        charucoCorners, charucoIds, _, _ = charuco_detector.detectBoard(image, markerCorners=markerCorners,
+                                                                        markerIds=markerIds)
+        self.assertEqual(len(charucoIds), 6)
+        self.assertEqual(len(markerIds), 6)
+
+        for i, id in enumerate(markerIds.reshape(-1)):
+            np.testing.assert_allclose(gold_corners[id], markerCorners[i].reshape(4, 2), 0.01, 1.)
+
     def test_write_read_dictionary(self):
         try:
             aruco_dict = cv.aruco.getPredefinedDictionary(cv.aruco.DICT_5X5_50)

--- a/modules/objdetect/src/aruco/aruco_detector.cpp
+++ b/modules/objdetect/src/aruco/aruco_detector.cpp
@@ -1000,7 +1000,13 @@ static inline void _projectUndetectedMarkers(const Board &board, InputOutputArra
                                              OutputArray undetectedMarkersIds) {
     Mat rvec, tvec; // first estimate board pose with the current avaible markers
     Mat objPoints, imgPoints; // object and image points for the solvePnP function
-    board.matchImagePoints(detectedCorners, detectedIds, objPoints, imgPoints);
+    // To refine corners of ArUco markers the function refineDetectedMarkers() find an aruco markers pose from 3D-2D point correspondences.
+    // To find 3D-2D point correspondences uses matchImagePoints().
+    // The method matchImagePoints() works with ArUco corners (in Board/GridBoard cases) or with ChArUco corners (in CharucoBoard case).
+    // To refine corners of ArUco markers we need work with ArUco corners only in all boards.
+    // To call matchImagePoints() with ArUco corners for all boards we need to call matchImagePoints() from base class Board.
+    // The method matchImagePoints() implemented in Pimpl and we need to create temp Board object to call the base method.
+    Board(board.getObjPoints(), board.getDictionary(), board.getIds()).matchImagePoints(detectedCorners, detectedIds, objPoints, imgPoints);
     if (objPoints.total() < 4ull) // at least one marker from board so rvec and tvec are valid
         return;
     solvePnP(objPoints, imgPoints, cameraMatrix, distCoeffs, rvec, tvec);

